File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Services/PanoAudioService.m
//
//  PanoAudioService.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/15.
//

#import "PanoAudioService.h"
#import "PanoClientService.h"
#import <AVFoundation/AVFoundation.h>

NSString *PanoAudioDumpFileName = @"pano_audio.dump";

UInt64 PanoMaxAudioDumpFileSize = 200 * 1024 * 1024;

@implementation PanoAudioService

- (instancetype)init
{
    self = [super init];
    if (self) {
        _isMuted = false;
    }
    return self;
}

- (UInt64)userId {
    return PanoClientService.service.config.userId;
}

- (void)muteAudio {
    _isMuted = true;
    [PanoClientService.service.engineKit muteAudio];
    [PanoClientService.service.userService onUserAudioMute:self.userId];
}

- (void)unmuteAudio {
    _isMuted = false;
    [PanoClientService.service.engineKit unmuteAudio];
    [PanoClientService.service.userService onUserAudioUnmute:self.userId];
}

- (void)toggleAudio {
    _isMuted ? [self unmuteAudio] : [self muteAudio];
}

- (void)startAudio:(BOOL)mute {
    [PanoClientService.service.engineKit startAudio];
    [PanoClientService.service.userService onUserAudioStart:self.userId];
    if (mute) {
        [self muteAudio];
    }
}


- (void)stopAudio {
    [PanoClientService.service.engineKit stopAudio];
    [PanoClientService.service.userService onUserAudioStop:self.userId];
}

- (BOOL)isEnabledLoudspeaker {
    return [PanoClientService.service.engineKit isEnabledLoudspeaker];
}

- (void)setLoudspeakerStatus:(BOOL)staus {
    [PanoClientService.service.engineKit setLoudspeakerStatus:staus];
}

- (void)startAudioDump:(NSTimeInterval)timeinterval {
    [NSObject cancelPreviousPerformRequestsWithTarget:self];
    NSString *panoAudioDumpFile = [NSTemporaryDirectory() stringByAppendingPathComponent:PanoAudioDumpFileName];
    [PanoClientService.service.engineKit startAudioDumpWithFilePath:panoAudioDumpFile maxFileSize:PanoMaxAudioDumpFileSize];
    if (timeinterval > 0) {
        [self performSelector:@selector(stopAudioDump) withObject:nil afterDelay:timeinterval];
    }
}

- (void)stopAudioDump {
    [PanoClientService.service.engineKit stopAudioDump];
}

- (void)enableAudioEarMonitoring:(BOOL)enable {
    [PanoClientService.service.engineKit setOption:@(enable) forType:kPanoOptionAudioEarMonitoring];
}

- (void)setAudioDeviceVolume:(UInt32)volume {
    [PanoClientService.service.engineKit setAudioDeviceVolume:volume withType:kPanoDeviceAudioRecord];
}

+ (BOOL)hasHeadphonesDevice {
    AVAudioSessionRouteDescription *route = AVAudioSession.sharedInstance.currentRoute;
    for (AVAudioSessionPortDescription *port in route.outputs) {
        if ([port.portType isEqualToString: AVAudioSessionPortHeadphones] ||
            [port.portType isEqualToString: AVAudioSessionPortBluetoothLE] ||
            [port.portType isEqualToString: AVAudioSessionPortBluetoothHFP] ||
            [port.portType isEqualToString: AVAudioSessionPortBluetoothA2DP]) {
            return true;
        }
    }
    return false;
}

@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Service/PanoAudioService.swift
//
//  PanoAudioService.swift
//  PanoAudioRoom
//
//  Created by wenhua.zhang on 2020/10/28.
//

import Foundation
import AVFoundation


let PanoAudioDumpFileName = "pano_audio.dump";

let PanoMaxAudioDumpFileSize: Int64 = 200 * 1024 * 1024;

protocol PanoAudioInterface {
    
    func unmuteAudio();
    
    func muteAudio();
    
    func startAudio(mute: Bool);
    
    func startAudio();
    
    func stopAudio();
    
    func startAudioDump(timeinterval: TimeInterval);
    
    func stopAudioDump();
}

protocol PanoAudioMixingInterface {
    
    func createAudioMixingTask(fileName: String);
    
    func destroyAudioMixingTask(fileName: String);
    
    func destroyAllAudioMixingTask()
    
    
}

class PanoAudioService : PanoBaseService  {
    
    override class func service() -> PanoAudioService? {
        return PanoServiceManager.service(type: .audio) as? PanoAudioService
    }
    
    var isMuted = false
    
    deinit {
        NSObject.cancelPreviousPerformRequests(withTarget: self)
    }
}


extension PanoAudioService: PanoAudioInterface {

    func muteAudio() {
        isMuted = true
        let result = PanoClientService.rtcEngineKit()?.muteAudio()
        print("muteAudio \(String(describing: result))")
        PanoUserService.service()?.onUserAudioMute(PanoClientService.service().userId)
    }
    
    func unmuteAudio() {
        isMuted = false
        let result = PanoClientService.rtcEngineKit()?.unmuteAudio()
        print("stopAudio \(String(describing: result))")
        PanoUserService.service()?.onUserAudioUnmute(PanoClientService.service().userId)
    }
    
    func toggleAudio() {
        isMuted ? unmuteAudio() : muteAudio()
    }
    
    func startAudio(mute: Bool) {
        startAudio()
        if mute {
            self.muteAudio()
        }
    }
    
    func startAudio() {
        let result = PanoClientService.rtcEngineKit()?.startAudio()
        print("startAudio \(String(describing: result))")
        PanoUserService.service()?.onUserAudioStart(PanoClientService.service().userId)
    }
    
    func stopAudio() {
        let result = PanoClientService.rtcEngineKit()?.stopAudio()
        print("stopAudio \(String(describing: result))")
        PanoUserService.service()?.onUserAudioStop(PanoClientService.service().userId)
    }
    
    func isEnabledLoudspeaker() -> Bool {
        if let engint = PanoClientService.rtcEngineKit() {
            return engint.isEnabledLoudspeaker()
        }
        return false
    }
    
    func setLoudspeakerStatus(staus: Bool) {
        PanoClientService.rtcEngineKit()?.setLoudspeakerStatus(staus)
    }
    
    func startAudioDump(timeinterval: TimeInterval = -1) {
        NSObject.cancelPreviousPerformRequests(withTarget: self)
        let PanoAudioDumpFile = NSTemporaryDirectory() + PanoAudioDumpFileName
        PanoClientService.rtcEngineKit()?.startAudioDump(withFilePath: PanoAudioDumpFile, maxFileSize: PanoMaxAudioDumpFileSize)
        if timeinterval > 0 {
            self.perform(#selector(stopAudioDump), with: nil, afterDelay: timeinterval)
        }
    }
    
    @objc func stopAudioDump() {
        PanoClientService.rtcEngineKit()?.stopAudioDump()
    }
    
    func enableAudioEarMonitoring(enable: Bool) {
        PanoClientService.rtcEngineKit()?.setOption(enable as NSObject, for: .optionAudioEarMonitoring)
    }
    
    func setAudioDeviceVolume(volume: UInt32) {
        PanoClientService.rtcEngineKit()?.setAudioDeviceVolume(volume, with: .audioRecord)
    }
    
    class func hasHeadphonesDevice() -> Bool {
        let route = AVAudioSession.sharedInstance().currentRoute
        for port in route.outputs {
            if  port.portType == AVAudioSession.Port.headphones ||
                port.portType == AVAudioSession.Port.bluetoothLE ||
                port.portType == AVAudioSession.Port.bluetoothHFP ||
                port.portType == AVAudioSession.Port.bluetoothA2DP {
                return true
            }
        }
        return false
    }
}
